### PR TITLE
CONTRIBUTING.md: Add section on reviewing PRs

### PR DESCRIPTION
**Pull request by ssiccha**

### CODE REVIEW

**Comment by ssiccha** [comment]
@markusbaumeister Could you also have a look? :)

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -141,6 +141,27 @@ to email the patch to <gap@gap-system.org>.  You can refer to
...
+
+### Before you even run the code (see [here](https://lornajane.net/posts/2015/code-reviews-before-you-even-run-the-code))
+
+* Is it clear what feature / fix the contribution adresses?
+* Does it relate to exactly one issue?
+* Do the changed files look sensible? Use `git log --stat` to discover
+possibly unintentional changes to files.
+* Do the commit messages look good? Should some commits be squashed / broken up?
+
+### Look at the code itself
```
Comment (markusbaumeister) [inline CONTRIBUTING.md]: I would add a bullet point here:
* Is the code easy to maintain? Are variable names sensibly chosen? Are there enough comments to easily understand the code?

Comment (ssiccha) [reply CONTRIBUTING.md]: sounds good. I'll make a few bullet points out of that point though

**Comment by ssiccha** [comment]
@markusbaumeister thx for the feedback.
I will squash the commits later.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -141,6 +141,33 @@ to email the patch to <gap@gap-system.org>.  You can refer to
 [this tutorial](https://burzalodowa.wordpress.com/2013/10/05/how-to-send-patches-with-git-send-email/)
 on how to do this.
 
+## Reviewing code contributions
+
+When reviewing code, you can use the following list as a checklist.
```
Comment (fingolfin) [inline CONTRIBUTING.md]: This text seems to know that readers already know what code reviews are (and what their ultimate purpose is), and how to do them. Might be worthwhile to elaborate a bit on that, and/or link to source explaining it. For the latter (how to), https://help.github.com/articles/about-pull-request-reviews/ might be a good URL.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -141,6 +141,33 @@ to email the patch to <gap@gap-system.org>.  You can refer to
...
+
+### Before you even run the code (see [here](https://lornajane.net/posts/2015/code-reviews-before-you-even-run-the-code))
+
+* Is it clear what feature / fix the contribution adresses?
+* Does it relate to exactly one issue?
+* Do the changed files look sensible? Use `git log --stat` to discover
+possibly unintentional changes to files.
```
Comment (fingolfin) [inline CONTRIBUTING.md]: indent this line to align with the list text above it

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -141,6 +141,33 @@ to email the patch to <gap@gap-system.org>.  You can refer to
 [this tutorial](https://burzalodowa.wordpress.com/2013/10/05/how-to-send-patches-with-git-send-email/)
 on how to do this.
 
+## Reviewing code contributions
+
+When reviewing code, you can use the following list as a checklist.
+
+### Before you even run the code (see [here](https://lornajane.net/posts/2015/code-reviews-before-you-even-run-the-code))
```
Comment (fingolfin) [inline CONTRIBUTING.md]: This seems to assume a code reviewer will necessarily run the code. But I actually almost never do.

BTW, reading this text, it could very well also be addressed at somebody who wants to contribute a PR, rather than review it... Which makes sense, I guess. Might be worthwhile to point this out somewhere. Or perhaps move (some of) this list into a "How to contribute code" section, and then here just say that a reviewer should check if the criteria listed in that section are met.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -141,6 +141,33 @@ to email the patch to <gap@gap-system.org>.  You can refer to
...
+
+### Look at the code itself
+
+* Is the code correct?
+* Is the code commented where necessary?
+* Does the new code fit in with documented behaviour?
+* Are new features documented if necessary?
+* Are there tests if necessary?
+* Double check whether the changes should be included into the release notes.
+If not, label the Issue / PR accordingly.
```
Comment (fingolfin) [inline CONTRIBUTING.md]: Indent this line. Also, "Issue" -> "issue"

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -141,6 +141,33 @@ to email the patch to <gap@gap-system.org>.  You can refer to
...
 
+## Reviewing code contributions
+
+When reviewing code, you can use the following list as a checklist.
+
+### Before you even run the code (see [here](https://lornajane.net/posts/2015/code-reviews-before-you-even-run-the-code))
+
+* Is it clear what feature / fix the contribution adresses?
+* Does it relate to exactly one issue?
+* Do the changed files look sensible? Use `git log --stat` to discover
```
Comment (fingolfin) [inline CONTRIBUTING.md]: I find the sentence "Do the changed files look sensible?" rather unclear, I am not sure what it really is meant to say. E.g. is it meant to mean "Do the changes overall look fine?" or "Does the *list of* changed files look sensible?" ?  The next sentence suggest the second meaning could be intended. I suggest to rephrase it suitably.

File: CONTRIBUTING.md (hunk 3 above)
Comment (fingolfin) [reply CONTRIBUTING.md]: Also, the above assumes that one has cloned the code locally. Which one can of course do -- but I assume many people will just review in the web interface, and there it is of course also possible to perform this check. So perhaps rephrase this to be neutral of the medium. E.g. 
> Does the list of changed files look sensible, or are there perhaps files with unintentional changes? On GitHub, do XYZ to find out; if you have clone the PR, you can use `git log --stat`.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -141,6 +141,33 @@ to email the patch to <gap@gap-system.org>.  You can refer to
...
+possibly unintentional changes to files.
+* Do the commit messages look good? Should some commits be squashed / broken up?
+* Eyeball the diff for
+  * Large commented / unused sections of code
+  * Strange variable or function names 
+  * Duplicate code
+
+
+### Look at the code itself
```
Comment (fingolfin) [inline CONTRIBUTING.md]: A lot of the previous section already deals with looking at the code itself, doesn't it? It's not so clear to me why this new sections starts exactly here...

File: CONTRIBUTING.md (hunk 4 above)
Comment (ssiccha) [reply CONTRIBUTING.md]: I'ld add this to the end of the `Making changes` section:
> Feel free to read the section [the reviewing process](https://github.com/gap-system/gap/blob/master/CONTRIBUTING.md#the-reviewing-process) to make the review of your pull request go smoothly.
What do you think?

**Comment by ssiccha** [comment]
Do you think this
> Please be careful to criticize constructively and not use dismissive language
is enough or should I elaborate more on that?
On the other hand, are the other points too detailed? Should I emphasize some points (e.g. make them bold) or leave some other points out entirely / move them into a "this is not so important" section?

**Comment by ssiccha** [comment]
https://alexgaynor.net/2013/sep/26/effective-code-review/ has veery helpful remarks. I'll redo the new section based on what he writes.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -126,6 +126,9 @@ From there you will have to wait on one of the GAP committers to respond to
 the request. This response might be an accept or some
 changes/improvements/alternatives will be suggested.  We do not guarantee
 that all requests will be accepted.
+Feel free to read the section
+[the reviewing process](https://github.com/gap-system/gap/blob/master/CONTRIBUTING.md#the-reviewing-process)
```
Comment (fingolfin) [inline CONTRIBUTING.md]: This hardcodes a link to the master branch, which is not ideal, as this might appear in other branches, and ultimately in a release version of GAP, where this makes even less sense. But I think this will just work, too: `[the reviewing process](#the-reviewing-process)`. Alternatively, just do without the link, and say:
> ... read the section "The reviewing process" later in this file ...

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -141,6 +144,57 @@ to email the patch to <gap@gap-system.org>.  You can refer to
 [this tutorial](https://burzalodowa.wordpress.com/2013/10/05/how-to-send-patches-with-git-send-email/)
 on how to do this.
 
+## The reviewing process
+
+Before a change can be incorporated into the code base, our setup requires
+each pull request (PR) to be
+[reviewed](https://help.github.com/articles/about-pull-request-reviews/)
+by at least one GAP developer (i.e. anybody with write access to the
+[gap repository](https://github.com/gap-system/gap/)).
```
Comment (fingolfin) [inline CONTRIBUTING.md]: Hmm "our setup requires" sounds as if it wasn't a conscious decision, but rather is an accident of technical requirement

And a "GAP developer" is anybody with write access... ? I am not sure I'd agree :-). Also, there does not seem a way to find out who has write access to a repository, so this is not all that helpful in practice either. How about something more like this:

> Before any change is incorporated into the GAP code base, it must undergo a mandatory code review.
> Typically, this is done for each pull request (PR) via the
> [GitHub code review facilitates]((https://help.github.com/articles/about-pull-request-reviews/)).
> The PR must have at least one approving code review from a core GAP developer with write access to
> the GAP code repository. However, everybody is very welcome to submit code reviews. This helps
> the core developers a lot, and is a step towards become one of them yourself.
>
> To do so, search ...

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -141,6 +144,57 @@ to email the patch to <gap@gap-system.org>.  You can refer to
...
+
+If you don't have write access, you are nevertheless very welcome to submit
+code reviews! Search the
+[pull requests](https://github.com/gap-system/gap/pulls)
+for a PR that you would like to review. Once you have chosen a PR
+(e.g. [PR 2261](https://github.com/gap-system/gap/pull/2261)),
+open the `Files changed` tab
+(e.g. [this](https://github.com/gap-system/gap/pull/2261/files))
+and add your comments or click the green `Review changes` button in the top
+right.
```
Comment (fingolfin) [inline CONTRIBUTING.md]: It's not really an "or" though, is it? In the end, you always must click the green button to submit your review.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -141,6 +144,57 @@ to email the patch to <gap@gap-system.org>.  You can refer to
...
+
+You can use the lists below as checklists for how to write your review.
+Please be careful to criticize constructively and not use dismissive language
+(see e.g. Brian Lee's section on `Rewording Feedback` in his
+[blog post](https://medium.com/unpacking-trunk-club/designing-awesome-code-reviews-5a0d9cd867e3)).
```
Comment (fingolfin) [inline CONTRIBUTING.md]: Fine

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -141,6 +144,57 @@ to email the patch to <gap@gap-system.org>.  You can refer to
...
+
+
+### Before you dive into the code (taken from [here](https://lornajane.net/posts/2015/code-reviews-before-you-even-run-the-code))
```
Comment (fingolfin) [inline CONTRIBUTING.md]: I'd move the link out of the title. E.g. write this as first sentence of this section:
> This section is based on [this article](...).

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -155,3 +209,5 @@ Heavily adapted from the contributing files from the
 [Puppet project](https://github.com/puppetlabs/puppet),
 [Factory Girl Rails](https://github.com/thoughtbot/factory_girl_rails/blob/master/CONTRIBUTING.md),
 and [Idris](https://github.com/idris-lang/Idris-dev).
+The section on reviewing code is partly based on
+[lornajane's blog entry](https://lornajane.net/posts/2015/code-reviews-before-you-even-run-the-code).
```
Comment (fingolfin) [inline CONTRIBUTING.md]: This would be redundant if you mention it in the section as suggested above, no?

File: CONTRIBUTING.md (hunk 9 above)
Comment (ssiccha) [reply CONTRIBUTING.md]: Ah, then that was my "obvious" definition which was only in my head and apparently also wrong. :)

File: CONTRIBUTING.md (hunk 10 above)
Comment (ssiccha) [reply CONTRIBUTING.md]: GitHub also adds a _User reviewed X days ago_ to the timeline if you only add comments. I think this would be equivalent to doing _Comment_ instead of _Approve_ or _Request Changes_.

Comment (fingolfin) [reply CONTRIBUTING.md]: That's not quite accurate, as is the whole description: If you clikc on a line to comment, you can either use the "Add single comment" button -- then a comment is added immediately, but no "User reviewed ... days ago" comment is added. Or you use the "Start a review" button, then added multiple comments; but these stay private until you use the "Review changes" button (and then in there, you can choose from three option, one being to just "comment").

But instead of describing this in detail here, I'd just link to https://help.github.com/articles/reviewing-proposed-changes-in-a-pull-request/ here. Also, IMHO it's better to avoid excessive use of the `[description](URL)` feature of Markdown and instead write links directly where it makes sense for the text flow -- this file is meant to be readable without a Markdown renderer, too, after all. So how about this:

> To review some code, start by browsing the list of open pull requests (PRs) at
>  <https://github.com/gap-system/gap/pulls> and look for a PR you would like
>  to review. Once you have chosen one, you can comment on its content, and
>  even individual lines changed by it, by following the instructions given on
> <https://help.github.com/articles/reviewing-proposed-changes-in-a-pull-request/>

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -126,6 +126,9 @@ From there you will have to wait on one of the GAP committers to respond to
 the request. This response might be an accept or some
 changes/improvements/alternatives will be suggested.  We do not guarantee
 that all requests will be accepted.
+Feel free to read the section
+[the reviewing process](#the-reviewing-process)
```
Comment (fingolfin) [inline CONTRIBUTING.md]: I would not use a hyperlink here. It makes the document more annoying to read without a Markdown renderer, and IMHO is not that useful with one either.

**Comment by ChrisJefferson** [comment]
I am happy with these changes, and think they are a significant improvement.

File: CONTRIBUTING.md (hunk 14 above)
Comment (ssiccha) [reply CONTRIBUTING.md]: Ok, that's a point I didn't consider.

File: CONTRIBUTING.md (hunk 10 above)
Comment (ssiccha) [reply CONTRIBUTING.md]: Ah, I didn't look for a github help page. Awesome!
